File: lib/data/models/article.dart
import 'package:equatable/equatable.dart';

import 'github_repo.dart';
import 'social_post.dart';

enum ArticleSourceType { news, blog, hackerNews, reddit, productHunt, github }

class ArticleSource extends Equatable {
  final String id;
  final String name;
  final String? iconUrl;
  final ArticleSourceType type;

  const ArticleSource({
    required this.id,
    required this.name,
    this.iconUrl,
    required this.type,
  });

  @override
  List<Object?> get props => [id, name, iconUrl, type];
}

class Article extends Equatable {
  final String id;
  final String title;
  final String? summary;
  final String? body;
  final String? imageUrl;
  final String url;
  final ArticleSource source;
  final String? author;
  final DateTime publishedAt;
  final List<String> topics;
  final List<String> companies;
  final List<String> stack;
  final int trendScore;
  final int viralityScore;
  final String? whyItMatters;
  final List<String> keyPoints;
  final List<GithubRepo> relatedRepos;
  final List<SocialPost> discussions;
  final bool isSaved;
  final bool isRead;

  const Article({
    required this.id,
    required this.title,
    this.summary,
    this.body,
    this.imageUrl,
    required this.url,
    required this.source,
    this.author,
    required this.publishedAt,
    this.topics = const [],
    this.companies = const [],
    this.stack = const [],
    this.trendScore = 0,
    this.viralityScore = 0,
    this.whyItMatters,
    this.keyPoints = const [],
    this.relatedRepos = const [],
    this.discussions = const [],
    this.isSaved = false,
    this.isRead = false,
  });

  Article copyWith({
    bool? isSaved,
    bool? isRead,
    String? summary,
    List<String>? keyPoints,
    String? whyItMatters,
    List<GithubRepo>? relatedRepos,
    List<SocialPost>? discussions,
  }) {
    return Article(
      id: id,
      title: title,
      summary: summary ?? this.summary,
      body: body,
      imageUrl: imageUrl,
      url: url,
      source: source,
      author: author,
      publishedAt: publishedAt,
      topics: topics,
      companies: companies,
      stack: stack,
      trendScore: trendScore,
      viralityScore: viralityScore,
      whyItMatters: whyItMatters ?? this.whyItMatters,
      keyPoints: keyPoints ?? this.keyPoints,
      relatedRepos: relatedRepos ?? this.relatedRepos,
      discussions: discussions ?? this.discussions,
      isSaved: isSaved ?? this.isSaved,
      isRead: isRead ?? this.isRead,
    );
  }

  @override
  List<Object?> get props => [id, isSaved, isRead];
}

File: backend/app/schemas/article.py
from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


class SourceType(str, Enum):
    news = "news"
    blog = "blog"
    hacker_news = "hacker_news"
    reddit = "reddit"
    product_hunt = "product_hunt"
    github = "github"


class ArticleSource(BaseModel):
    id: str
    name: str
    icon_url: Optional[str] = None
    type: SourceType


class GithubRepo(BaseModel):
    id: str
    full_name: str
    description: str
    url: str
    language: Optional[str] = None
    stars: int = 0
    forks: int = 0
    stars_this_week: int = 0
    readme_excerpt: Optional[str] = None
    last_commit: Optional[datetime] = None


class SocialPlatform(str, Enum):
    reddit = "reddit"
    twitter = "twitter"
    hacker_news = "hacker_news"


class SocialPost(BaseModel):
    id: str
    platform: SocialPlatform
    author: str
    author_handle: Optional[str] = None
    content: str
    url: str
    upvotes: int = 0
    comments: int = 0
    posted_at: datetime
    sentiment: str = "neutral"


class Article(BaseModel):
    id: str
    title: str
    summary: Optional[str] = None
    body: Optional[str] = None
    image_url: Optional[str] = None
    url: str
    source: ArticleSource
    author: Optional[str] = None
    published_at: datetime
    topics: List[str] = Field(default_factory=list)
    companies: List[str] = Field(default_factory=list)
    stack: List[str] = Field(default_factory=list)
    trend_score: int = 0
    virality_score: int = 0
    why_it_matters: Optional[str] = None
    key_points: List[str] = Field(default_factory=list)
    related_repos: List[GithubRepo] = Field(default_factory=list)
    discussions: List[SocialPost] = Field(default_factory=list)


class FeedResponse(BaseModel):
    items: List[Article]
    next_page: Optional[int] = None
